From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Sun, 2 Jun 2019 00:02:39 +0200
Subject: [PATCH] {MAYBE UPS} scsi/fcoe: silence -Warray-bounds

gcc 9 does not like calling memset on the first member of a structure to
zeroize the whole structure:

    In function 'memset',
        inlined from 'fcoe_ctlr_vlan_parse' at drivers/scsi/fcoe/fcoe_ctlr.c:2830:2,
        inlined from 'fcoe_ctlr_vlan_recv' at drivers/scsi/fcoe/fcoe_ctlr.c:3005:7:
    include/linux/string.h:356:9: error: '__builtin_memset' offset [321,
    344] from the object at 'buf' is out of the bounds of referenced
    subobject 'rdata' with type 'struct fc_rport_priv' at offset 0
    [-Werror=array-bounds]
      356 |  return __builtin_memset(p, c, size);
          |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
---
 drivers/scsi/fcoe/fcoe_ctlr.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/scsi/fcoe/fcoe_ctlr.c b/drivers/scsi/fcoe/fcoe_ctlr.c
index 590ec8009f52..4f68482fd071 100644
--- a/drivers/scsi/fcoe/fcoe_ctlr.c
+++ b/drivers/scsi/fcoe/fcoe_ctlr.c
@@ -2269,8 +2269,9 @@ static void fcoe_ctlr_vn_start(struct fcoe_ctlr *fip)
  */
 static int fcoe_ctlr_vn_parse(struct fcoe_ctlr *fip,
 			      struct sk_buff *skb,
-			      struct fc_rport_priv *rdata)
+			      void *arg_rdata)
 {
+	struct fc_rport_priv *rdata = arg_rdata;
 	struct fip_header *fiph;
 	struct fip_desc *desc = NULL;
 	struct fip_mac_desc *macd = NULL;
@@ -2284,7 +2285,7 @@ static int fcoe_ctlr_vn_parse(struct fcoe_ctlr *fip,
 	u32 dtype;
 	u8 sub;
 
-	memset(rdata, 0, sizeof(*rdata) + sizeof(*frport));
+	memset(arg_rdata, 0, sizeof(*rdata) + sizeof(*frport));
 	frport = fcoe_ctlr_rport(rdata);
 
 	fiph = (struct fip_header *)skb->data;
@@ -2757,7 +2758,7 @@ static int fcoe_ctlr_vn_recv(struct fcoe_ctlr *fip, struct sk_buff *skb)
 		goto drop;
 	}
 
-	rc = fcoe_ctlr_vn_parse(fip, skb, &buf.rdata);
+	rc = fcoe_ctlr_vn_parse(fip, skb, &buf);
 	if (rc) {
 		LIBFCOE_FIP_DBG(fip, "vn_recv vn_parse error %d\n", rc);
 		goto drop;
@@ -2802,8 +2803,9 @@ static int fcoe_ctlr_vn_recv(struct fcoe_ctlr *fip, struct sk_buff *skb)
  */
 static int fcoe_ctlr_vlan_parse(struct fcoe_ctlr *fip,
 			      struct sk_buff *skb,
-			      struct fc_rport_priv *rdata)
+			      void *arg_rdata)
 {
+	struct fc_rport_priv *rdata = arg_rdata;
 	struct fip_header *fiph;
 	struct fip_desc *desc = NULL;
 	struct fip_mac_desc *macd = NULL;
@@ -2815,7 +2817,7 @@ static int fcoe_ctlr_vlan_parse(struct fcoe_ctlr *fip,
 	u32 dtype;
 	u8 sub;
 
-	memset(rdata, 0, sizeof(*rdata) + sizeof(*frport));
+	memset(arg_rdata, 0, sizeof(*rdata) + sizeof(*frport));
 	frport = fcoe_ctlr_rport(rdata);
 
 	fiph = (struct fip_header *)skb->data;
@@ -2990,7 +2992,7 @@ static int fcoe_ctlr_vlan_recv(struct fcoe_ctlr *fip, struct sk_buff *skb)
 
 	fiph = (struct fip_header *)skb->data;
 	sub = fiph->fip_subcode;
-	rc = fcoe_ctlr_vlan_parse(fip, skb, &buf.rdata);
+	rc = fcoe_ctlr_vlan_parse(fip, skb, &buf);
 	if (rc) {
 		LIBFCOE_FIP_DBG(fip, "vlan_recv vlan_parse error %d\n", rc);
 		goto drop;
-- 
